FrameworkSystem/DB/ProxyDB.py

```diff
@@ -623,6 +773,11 @@ def getProxy(self, userDN, userGroup, requiredLifeTime=False):
 
     # Standard proxy is requested
     retVal = self.__getPemAndTimeLeft(userDN, userGroup)
+    if retVal['OK'] and requiredLifeTime:
+      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
+        retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
+    else:
+      retVal = self.__generateProxy(userDN, userGroup, requiredLifeTime=requiredLifeTime)
```

Comment:
So, `if retVal['Value'][1] >= requiredLifeTime` (the if above is not entered), but I don't have proxyProviders, this will return an error, and thus this whole function will return S_ERROR?

Reply:
In this case retVal['OK'] is true, so block (if not retVal['OK']: return retVal) will be ignore

Reply:
Right, I messed up which `if` this `else` belongs to, the problem actually happens when `requiredLifeTime == False`, then even though `retVal['OK']` is true one ends up in this `else`, or am I again missing something?

Reply:
Code in last redaction:
# retVal['OK'] == True & requiredLifeTime == False

if retVal['OK'] and requiredLifeTime:
  # in this case we ignore it
  if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
    errMsg = 'the required lifetime is less than the time left in the proxy'
    retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)

else:
  # and come here
  errMsg = retVal['Message']
  # where we try to generate proxy with requiredLifeTime=False
  retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)

if retVal['OK']:
  # all ok, continue

I do not see the problem, maybe I do not understand what you mean

Reply:
https://github.com/TaykYoku/DIRAC/blob/185a34d41d645d7bcb4ed6fd815b0ce2e8531e96/FrameworkSystem/DB/ProxyDB.py#L766-L773

```python
    requiredLifeTime  = False # From the function argument
    retVal = self.__getPemAndTimeLeft(userDN, userGroup)  # returns S_OK
    if retVal['OK'] and requiredLifeTime: # is not entered
      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
        errMsg = 'the required lifetime is less than the time left in the proxy'
        retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    else: # we end up here
      errMsg = retVal['Message'] # exception KeyError (retVal is S_OK(), doesn't have 'Message', but this is a different problem now)
      retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime) 
      # if exception before doesn't happen, this now returns S_ERROR, because
      # I don't have ProxyProviders defined, retVal is now S_ERROR instead of S_OK
```
One has to assume that there are cases out there where this function was previously called with requiredLifeTime evaluating to False, which would not result in an error (otherwise we can just return S_ERROR at the top...)

Reply:
In this case, we don`t have a proxy in the DB and cannot use MyProxy and cannot generate a new one, so the method just ended in an error that should catch the last line:
https://github.com/TaykYoku/DIRAC/blob/185a34d41d645d7bcb4ed6fd815b0ce2e8531e96/FrameworkSystem/DB/ProxyDB.py#L789-L790
```
    return S_ERROR("Can't get a proxy%s %s" % (requiredLifeTime and ' for %s seconds:' % requiredLifeTime or ':', errMsg))
```

Reply:
o, I seem to understand what you mean

Reply:
This logic will be correct

```
    retVal = self.__getPemAndTimeLeft(userDN, userGroup)
    if not retVal['OK']:
      errMsg = retVal['Message']
      retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    elif requiredLifeTime:
      if retVal['Value'][1] < requiredLifeTime and not self.__useMyProxy:
        errMsg = 'the required lifetime is less than the time left in the proxy'
        retVal = self.__generateProxyFromProxyProviders(userDN, userGroup, requiredLifeTime=requiredLifeTime)
    if retVal['OK']:
```

Reply:
Yes, this seems to not turn S_OK to S_ERROR. I will have another look once you pushed your changes.